gulpfile.js

```diff
@@ -124,28 +172,197 @@ gulp.task("default", function (callback) {
+                }
+                iOSEmulatorReady(iOSEmulatorReadyLooper);
+            });
+        }, true);
+    }
+    if (!restartIfRunning) {
+        if (android) {
+            androidEmulatorReady(() => {
+                killThenStartAndroid();
+            });
+        }
+        if (ios) {
+            iOSEmulatorReady(() => {
+                killThenStartIOS();
+            });
+        }
+    } else {
+        if (android) killThenStartAndroid();
+        if (ios) killThenStartIOS();
+    }
+    // This needs to be done so that the task will exit.
+    // The command that creates the Android emulator persists with the life of the emulator and hangs this process unless we force it to quit.
+    gulp.doneCallback = (err) => {
+        process.exit(err ? 1 : 0);
+    }
+}
 
-gulp.task("test-ios-uiwebview", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://127.0.0.1:3000", "--platform", "ios", "--use-wkwebview", "false"];
-    executeCommand(command, args, callback);
-});
+// procedurally generate tasks for every possible testing configuration
+var cleanSuffix = "-clean";
+var fastSuffix = "-fast";
 
-gulp.task("test-ios-wkwebview", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://127.0.0.1:3000", "--platform", "ios", "--use-wkwebview", "true"];
-    executeCommand(command, args, callback);
-});
+// generate tasks for starting emulators
+function generateEmulatorTasks(taskName, android, ios) {
+    gulp.task(taskName, function (callback) {
+        startEmulators(callback, false, android, ios);
+    });
+    
+    gulp.task(taskName + cleanSuffix, function (callback) {
+        startEmulators(callback, true, android, ios);
+    });
+}
 
-gulp.task("test-android", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://10.0.2.2:3000", "--platform", "android"];
-    executeCommand(command, args, callback);
-});
+function getEmulatorTaskNameSuffix(android, ios) {
+    var emulatorTaskNameSuffix = "";
+    
+    if (android) emulatorTaskNameSuffix += "-android";
+    if (ios) emulatorTaskNameSuffix += "-ios";
+    // "emulator" instead of "emulator-android-ios"
+    if (android && ios) emulatorTaskNameSuffix = "";
+    
+    return emulatorTaskNameSuffix
+}
 
-gulp.task("test", function (callback) {
-    runSequence("default", "test-android", "test-ios", callback);
-});
\ No newline at end of file
+var emulatorTaskNamePrefix = "emulator";
+for (var android = 0; android < 2; android++) {
+    for (var ios = 0; ios < 4; ios++) {
+        generateEmulatorTasks(emulatorTaskNamePrefix + getEmulatorTaskNameSuffix(android, ios), android, ios);
+    }
+}
+                
+function generateTestTasks(taskName, options) {
+    gulp.task(taskName + "-fast", function (callback) {
```

Comment:
For clarity, I think it's best to keep gulp tasks declared as globals and not generated at runtime.